### PR TITLE
feat: enhance types

### LINKED ISSUE #17: Improving the JSDoc types and generating a declaration file
I am working to throughly type `@wordpress/blocks` and `hpq` is a dependency of that package. I wanted to open the discussion on how best to improve the types produced by `hpq`.

There are a few options:

- Improve the types in the current JSDocs and generate a declaration file using TypeScript
- Move the package to TypeScript
- Publish a type declaration to DefinitelyTyped. This is my least favorite option, splitting the types from the source creates a difficult to maintain system where the types can get out of sync with the package.

I can do the work to integrate this, but I want to make sure there is interest before working on it.

### PR DESCRIPTION
**Pull request by johnhooks**
## Goal

Enhance JSDoc types to be more explicit, enable TypeScript checking of JavaScript files, and prepare for publishing first-party types.

Closes #17

### Type output directory

The types are currently output to the `es` folder. I thought that was more appropriate than the `dist` folder with the `UMD` builds, though I can change it to whatever you think is best.

### JSDoc parameter descriptors `?` or `=`

TypeScript interprets `?` as a union with `null` and `=` as a union  with `undefined` in all the instances I believe the correct union should be `undefined`, that is why I change it.

### Using JSDoc over TS

Generics in JSDoc are awkward and I am not familiar with the syntax. To figure out the solution, I first converted `src/index.js` to a TypeScript file and worked in what I was most familiar with. I wasn't able to bring all the same types into the JSDoc version.

### CODE REVIEW

**Comment by johnhooks** [comment]
@aduth checkout this file, I typed everything like this and then ported it back into `src/index.js`. I think it is better typed, though to integrate TS files into the project needs more work. I just wanted to offer this up as another solution.

https://github.com/johnhooks/hpq/blob/172fec770524b0b10ed5a2d3e0ff1fd427740633/src/index.ts

**Comment by aduth** [comment]
Thanks for the quick turnaround on the pull request @johnhooks !

I might not be able to take a deep dive on this until the weekend, and shamefully I have to admit that it's been a while since the last major update to this library, so I'm having to refamiliarize myself with it as well!

A couple quick comments to what you've surfaced:

- Looking at the overloading in the TypeScript version, I think I'd be open to porting this to TypeScript if it'd make it easier for you to type it
- To your question about the return type of `MatcherFn`, I think conceptually I had imagined it _could_ return anything in the sense that someone could provide their own matcher function that returns a boolean, for example. However, given the set of default matchers, I think it'd be fair to say `string` or `undefined` would be the standard expected return type, if it helps avoid the `any` type.
   - I'd wonder if there's any way to still support someone who might want to define their own matcher which has an arbitrary return type? Maybe a generic and inferred [`ReturnType`](https://www.typescriptlang.org/docs/handbook/utility-types.html#returntypetype)?
- I think you're correct that I had mistakenly mixed up `?` and `=` in assigning the "optional" type
- To your question about output directory, I think it makes sense as you've proposed, though it's also probably time for a rethink as far as what outputs the library supports broadly speaking (i.e. standardizing on ESM)

**Comment by johnhooks** [comment]
@aduth thanks! I'm excited to help. 

> Looking at the overloading in the TypeScript version, I think I'd be open to porting this to TypeScript if it'd make it easier for you to type it

The library seems simple, though it is actually hard to know exactly what is being accessed when the matchers are defined. I am willing to take some time and get it right. And would be happy to help maintain a TS version.

> I'd wonder if there's any way to still support someone who might want to define their own matcher which has an arbitrary return type? Maybe a generic and inferred [ReturnType](https://www.typescriptlang.org/docs/handbook/utility-types.html#returntypetype)?

I agree, there should be a way to infer the return type. It might take a little refactoring to make it work.

> To your question about output directory, I think it makes sense as you've proposed, though it's also probably time for a rethink as far as what outputs the library supports broadly speaking (i.e. standardizing on ESM)

I for one would support ESM only, that shouldn't affect the largest dependent of this package, Gutenberg. Using `tsc` build output for modules and types would simplify the build to a single step.

It's really cool that this little library is used deep down in the core of Gutenberg. And it's obviously doing a great job because it was able to just sit for two years without any issues!

File: src/index.js
Comment (aduth) [reply src/index.js:27]: These look good to me as-is 👍

**Comment by aduth** [comment]
Was looking if there might be a way to infer the return type based on the input. Definitely challenging with the overloaded signature. This seems to look okay at a glance, though unsure yet how well it could be ported back to JSDoc:

https://www.typescriptlang.org/play?#code/C4TwDgpgBAsghsAxgCwgJwGIDsA8GB8UAvFABRYD2AJhAFxQCiANhALYRbACUxhcWIANwAoUJFgIU6APIAjAFZ5CJAEoREFNFRwBnYGgCWWAOYAaCUlSZcBfCOEAzAK5ZEwAxSxQwcNDoh4UBAAHsAcVDoWUtY4-CD45tJBoeGR8JYyCkr45NR0jCzsnOasklY69NJc9ADeUADaANJQRlAA1hAgFA5Q0gC69GrATmhYACrgAdJNfYQAviLOru6e3r7+gSFhWBFRVtixAvg5lDT0zGwcwCVl6BVQGNVQQyPjk0qLLm4eXj5+ARhkttdulogc4glekDUntMopbLkzgVLsUoKUMn5KlAAD4PHg1YRQIlQNAQYajKA1BbCObCYQaLB6EkQHROJjAYhrf6kKgURBOIrAAB0smoIHMdQcFAo9EREB4REIpwgQrCoQAwp5thy5lwRAymaTWeyAEycv7+Hl8gVXEVi8ykUVUEAKwhOkBCoxYdAACTGMAAMnqgA

**Comment by johnhooks** [comment]
That looks really good to me.

From what I learned the only way to "overload" the function in JSDocs is a union type, and that would look pretty messy and require `const parser = function() {}` syntax to type it as a union of multiple functions.

Do you want me to submit a PR with the other overloads I created and you commit the parse function signature? And we convert this to TypeScript ? It does seem to be a better fit.

**Comment by aduth** [comment]
Apparently the just-released TypeScript 5.0 includes an `@overload` directive for JSDoc, which sounds like it might be the perfect fit here.

https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#overload-support-in-jsdoc

**Comment by johnhooks** [comment]
That sound awesome! I'll read up on it.

**Comment by johnhooks** [comment]
That just came out, I skimmed through it two days ago and didn't catch `@overload`, so many great improvements!

**Comment by aduth** [comment]
This seems to be working okay for me in testing locally, want me to go ahead and push it up? Or do you want to take a closer look at it?

```js

/**
 * @template {(node: Element) => any} F
 * @typedef {(node: Element) => any} MatcherFn
 */

/**
 * @template {(node: Element) => any} F
 * @typedef {Record<string, MatcherFn<F>>} MatcherObj
 */

/**
 * @template {MatcherObj<F>} O
 * @template {MatcherFn<any>} F
 *
 * @overload
 * @param {string|Element} source Source content
 * @param {O} matchers Object of matchers
 * @return {{[K in keyof O]: ReturnType<O[K]>}} Matched values, shaped by object
 */

/**
 * @template {MatcherFn<any>} F
 *
 * @overload
 * @param {string|Element} source Source content
 * @param {F} matcher Matcher function
 * @return {ReturnType<F>} Matched value
 */

/**
 * Given a markup string or DOM element, creates an object aligning with the
 * shape of the matchers object, or the value returned by the matcher.
 *
 * @template {MatcherObj<F>} O
 * @template {MatcherFn<any>} F
 *
 * @param {string|Element} source Source content
 * @param {O|F} matchers Matcher function or object of matchers
 */
export function parse(source, matchers) {
  // ...
}
```

**Comment by johnhooks** [comment]
I just pushed something up. Sorry I can walk it back. Can I also do the comments like that rather than all spaced out like they do on Gutenberg?

**Comment by johnhooks** [comment]
@aduth I was going to walk some of it back, but it was breaking everything else. Could you take a look at the current version, and push any updates. I'll stop working on it until I hear from you.

Also there are still some build issues with the generics, see if you cane make sense of it.

File: src/index.js
Comment (johnhooks) [inline src/index.js:51]: I meant to remove `any`

File: src/index.js
Comment (johnhooks) [inline src/index.js:66]: This should be `Matcher<T>`

**Comment by aduth** [comment]
Hey, no worries and no need to walk anything back. I had to step away for a bit anyways, so it worked out. And it looks like you've got most everything in place now. Glad that new feature worked well!

**Comment by johnhooks** [comment]
Only issue I see with the new feature is that non of the comments came with the overloads, in the output `es/index.d.ts`

**Comment by aduth** [comment]
And yes, I'd be fine to collapse the whitespace in the JSDoc if you want to do that.

**Comment by aduth** [comment]
> Only issue I see with the new feature is that none of the comments came with the overloads, in the output `es/index.d.ts`

Hm, yeah, I see what you mean 🤔 Looks like it could be partly related to https://github.com/microsoft/TypeScript/issues/27981 ? Not sure if there's a syntax variation to get those comments to "stick" into the output.

**Comment by johnhooks** [comment]
@aduth I tried adding the description to all the overloads, but nothing worked. Other than that, I think it's ready for review.

File: src/index.ts
Comment (aduth) [inline src/index.ts:1]: Did you intend to add this file?

**Comment by aduth** [comment]
> @aduth I tried adding the description to all the overloads, but nothing worked. Other than that, I think it's ready for review.

Yeah, I didn't have any luck either. Do you think it should be a blocker? I'm okay to ship it without the comments if you are. Seems like it might be a bug in TypeScript itself. Only other option I could conceive is to port to TypeScript.

**Comment by johnhooks** [comment]
@aduth This is a half measure solution, but I don't think there will need to be many updates to this, if your publishing from your computer you could manually add the comments. Not ideal but might be better than no comments.

**Comment by johnhooks** [comment]
Also, the `.ts` file I created is still in there. Removing it now.

**Comment by johnhooks** [comment]
Or if you like the idea of TS, I can support that. Either way, I think we've got the types nailed.

**Comment by aduth** [comment]
I'm not as big a fan of manually editing the `.d.ts` file prior to publish, since I'll definitely forget to do that the next time I need to publish an update. Another option could be to check in the `.d.ts` file to Git, using the TypeScript build as a starting point but adding revisions of the comments? Also not opposed to TypeScript itself.

**Comment by johnhooks** [comment]
I totally agree about editing build assets. I support checking it into git or TS, your call.

**Comment by johnhooks** [comment]
This issue with TS is ESLint would need a new configuration. I have experience with it.

**Comment by aduth** [comment]
I think checking in the `.d.ts` file might be the lowest lift option, and we could revisit TypeScript if/when it becomes a pain to maintain that file separately. If it is a bug in TypeScript, maybe it'd be fixed upstream by that point anyways.

**Comment by aduth** [comment]
I have a [personal standard ESLint configuration](https://www.npmjs.com/package/@aduth/eslint-config) which supports TypeScript I use in my other projects, if we end up going the TypeScript route. I just hadn't got around to implementing it in this project yet.

**Comment by aduth** [comment]
@johnhooks I pushed up changes in 60a48ca to check-in the `.d.ts` file. I also created an issue on the TypeScript repository at https://github.com/microsoft/TypeScript/issues/53350 in case it helps toward a resolution. Does this look okay to you?

One other thing that comes to mind is that the matcher object form supports arbitrary nesting of objects. Would it be easy enough to change the `MatcherObj` type from `Record<string, MatcherFn<T>>` to `Record<string, MatcherObj<T> | MatcherFn<T>>`? This is another behavior that unfortunately didn't have good test coverage 😕 , but works all the same.

[Playground](https://aduth.github.io/hpq/) example:

image

**Comment by johnhooks** [comment]
@aduth That issue filed on TypeScript looks great to me.

That type makes sense to me though the type checker isn't happy with it.

```js
/**
 * @typedef {(node: Element) => T} MatcherFn
 * @template T
 */

/**
 * @typedef {Record<string, MatcherObj<T> | MatcherFn<T>>} MatcherObj
 * @template T
 */

/**
 * @typedef {(MatcherFn<T>|MatcherObj<T>)} Matcher
 * @template T
 */
```

Results in:

```
Type alias 'MatcherObj' circularly references itself.ts(2456)
```

I have always had trouble understanding the nuance of circular references, sometimes they work fine.

This works though...

```
/**
 * @typedef {(node: Element) => T} MatcherFn
 * @template T
 */

/**
 * @typedef {{[ x: string ]: MatcherObj<T> | MatcherFn<T>} } MatcherObj
 * @template T
 */

/**
 * @typedef {(MatcherFn<T>|MatcherObj<T>)} Matcher
 * @template T
 */
```

**Comment by johnhooks** [comment]
Though that creates a new type error in parse.

**Comment by aduth** [comment]
I think I got it working in TypeScript with a combination of your suggestion of the object index signature, plus an update on the return signature of the object overload to be conditional based on the object value.

I couldn't quite get it working in JSDoc, though I did make a bit of progress by making the `@template` tags type-constrained like in [my earlier comment](https://github.com/aduth/hpq/pull/18#issuecomment-1474942934).

Maybe we're back at reconsidering TypeScript 😅 

https://www.typescriptlang.org/play?#code/C4TwDgpgBAsghsAxgCwgJwGIDsA8GB8UAvFABRYD2AJhAFxQCiANhALYRbACUxhcWIANwAoUJFgIU6APIAjAFZ5CJAN5QA2gA96AZ2BoAllgDmAXXrwkqNHMUEoAHwlX02JVAC+w4QDMArliIwAYUWFBgcGg6EHhQEJrAHFQ6zlKYuPwg+AA0UNJxCUkplmm2Svjk1HSMLOycuayS1jr00lz0KuoA0lBGUADWECAUPnnmed2mBYlYyVAYUAD8UABKEMB+aFgAKuAx0pOErZMeIv6BwaHhkdGx8TNzJdZumfgVlDT0zGwcwA1N6Ba83aq3Wmx2eyUZwCQRCYQiURiC3uRVSzwyAhyeWmqKeMgU5UqnxqP3qUEaLiirUcwKgKmEUEZUDQYK2dNOwi8wkQoT0zIgOj8TGAxGuiNIVAoiD8dWAADpZNQQLk1D4KBQOlBZJF6ESIDwiIQPhA5YkEgBhUIzEUeTxcEQ8rB8lmC4UAJlFCOiEqlMt+CqVuVIiqoIANhBDIDlRiw6AAEtsYAAZe1AA

**Comment by aduth** [comment]
@johnhooks Looks like we both had the same idea this afternoon, I had also gone down this path of porting in 489bcdf, but it looks like you beat me by a few minutes. 😄

**Comment by johnhooks** [comment]
Check it out, the only thing that I couldn't figure out was how to run the tests.

**Comment by johnhooks** [comment]
The overloads have all the comments now, like they should.

**Comment by johnhooks** [comment]
It looks like you know how to register babel with Ts extensions! Awesome.

**Comment by johnhooks** [comment]
I'm going to pause on this. There are still some issues to resolve, but it seemed like I might have been stepping on your toes! I'll wait to hear from you. Also, it seemed like you were getting in to it, do you want to implement the same for rememo and memize, or would you like help?

File: src/index.ts
Comment (johnhooks) [inline src/index.ts:217]: This right here is amazing @aduth, it is actually typed as a `NamedNodeMap|undefined` without any type assertions! I had a big smile when I deleted the assertion and the type held, what a good feeling! 🎉

**Review by johnhooks — COMMENTED** [review]
@aduth this PR is ready for a review. I'm very happy with the solutions we came up with. Thank you so much for your help and support.

**Review by aduth — APPROVED** [review]
This is all looking fantastic! Thanks for bearing with me through all the iterations. I think it turned out great in the end. I'll plan to get this merged and published either later today or at least by the end of the weekend.

File: src/index.ts
Comment (johnhooks) [reply src/index.ts:171]: @aduth would you like me to make these edits?

Comment (aduth) [reply src/index.ts:171]: Sure, if you agree, please do!

Comment (johnhooks) [reply src/index.ts:171]: Will do, I agree with all your suggestions.

**Review by johnhooks — COMMENTED** [review]
@aduth, it is ready for final review.

**Review by aduth — APPROVED** [review]
Awesome! Thanks again for all your work with this. Excited to see it help with your work typing Gutenberg packages. I'll be merging and publishing shortly.

**Comment by aduth** [comment]
This is now published as `hpq@1.4.0` 🚀

**Pull request merged**
